core/src/main/scala/org/apache/spark/Partitioner.scala

```diff
@@ -56,9 +56,10 @@ object Partitioner {
    */
   def defaultPartitioner(rdd: RDD[_], others: RDD[_]*): Partitioner = {
     val bySize = (Seq(rdd) ++ others).sortBy(_.partitions.size).reverse
-    for (r <- bySize if r.partitioner.isDefined) {
-      return r.partitioner.get
-    }
+    for {
+      r <- bySize
+      partitioner <- r.partitioner if r.partitions.nonEmpty
+    } return partitioner
```

Comment:
This is difficult to read. It took me a while to understand what it's doing. Would you remind rewriting this:
```
val part = bySize.find(_.partitions.nonEmpty).map(_.partitioner)
if (part.isDefined) {
  return part.get
}
```